### PR TITLE
Update process tests

### PR DESCRIPTION
**Pull request by pallavit**
This change updates the System.Diagnostics.Process tests

### CODE REVIEW

File: src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs [hunk 1]
```diff
@@ -7,45 +8,44 @@ class Program
...
-                // We're testing STDERR streams, not the JIT debugger. 
-                // This makes the process behave just like a crashing .NET app, but without the WER invocation
-                // nor the blocking dialog that comes with it, or the need to suppress that.
-                Console.Error.WriteLine(string.Format("Unhandled Exception: {0}", toLog.Message));
-                return 1;
             }
+            return 100;
+        }
 
+        public static void DelayTask()
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs]: What's the purpose of this method?  How does it differ from Thread.Sleep(100)?

File: src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs [hunk 2]
```diff
@@ -7,45 +8,44 @@ class Program
...
-                    }
+                    // To avoid potential issues with orphaned processes (say, the test exits before
+                    // exiting the process), we'll say "infinite" is actually 30 seconds.
+                    Thread.Sleep(30 * 1000);
                 }
-                else
+                else if (args[0].Equals("error"))
+                {
+                    Console.Error.WriteLine("ProcessTest_ConsoleApp.exe error stream");
+                    DelayTask();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs]: What's the purpose of these DelayTask() calls?  Seems like it's sleeping for 100ms, and if so, is that setting us up for lab failures if the timing is important?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 3]
```diff
@@ -3,580 +3,381 @@
...
         {
-            try
-            {
-                foreach (Process p in Process.GetProcessesByName(s_ProcessName.Replace(".exe", "")))
-                {
-                    p.Kill();
-                }
-            }
-            catch { }; // do nothing 
+            s_process = CreateProcessInfinite();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: The "infinite" process only runs for 30 seconds.  If we're using one such process for the whole ProcessTest suite, is there any concern that the process tests could end up taking longer than that on a given run and end up failing as a result?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 4]
```diff
@@ -3,580 +3,381 @@
...
     {
-        static ProcessTest()
-        {
-            TryKillExistingTestProcesses();
-        }
 
-        private static void TryKillExistingTestProcesses()
+        private const string s_ProcessName = "ProcessTest_ConsoleApp.exe";
+        private static bool s_Process_EnableRaiseEvents_isExitedEventHandlerCalled = false;
+        private Process s_process;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: s_process is an instance field; it shouldn't have the 's' prefix.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 5]
```diff
@@ -3,580 +3,381 @@
...
-    public partial class ProcessTest
+    public partial class ProcessTest : IDisposable
     {
-        static ProcessTest()
-        {
-            TryKillExistingTestProcesses();
-        }
 
-        private static void TryKillExistingTestProcesses()
+        private const string s_ProcessName = "ProcessTest_ConsoleApp.exe";
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: As a constant; this should just be "ProcessName" rather than "s_ProcessName".

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 6]
```diff
@@ -3,580 +3,381 @@
...
+
+            //SetAndCheckBasePriority(ProcessPriorityClass.RealTime, 24);
+            SetAndCheckBasePriority(ProcessPriorityClass.High, 13);
+            SetAndCheckBasePriority(ProcessPriorityClass.Idle, 4);
+            SetAndCheckBasePriority(ProcessPriorityClass.Normal, 8);
         }
 
-        [Fact, ActiveIssue(541)]
-        public static void Process_BasePriority()
+        public void DelayTask(double delayInMs)
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Same question as the previous DelayTask.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 7]
```diff
@@ -797,5 +598,21 @@ public static void Process_Environment()
             }
             );
         }
+
+        public void Dispose()
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: I'd suggest moving this up next to the ctor.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 8]
```diff
@@ -3,580 +3,381 @@
...
 
         [Fact]
-        public static void Process_MainModule()
+        public void Process_MainModule()
         {
             // Get MainModule property from a Process object
-            Process p = CreateProcessInfinite();
-            p.Start();
-            try
+            string moduleName = s_process.MainModule.ModuleName;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: The open issue related to MainModule shows that ModuleName can internally throw a null reference exception.  MainModule looks like this today:
```
ModuleInfo module = NtProcessManager.GetFirstModuleInfo(_processId);
return new ProcessModule(module);
```
If GetFirstModuleInfo returns null (which is possible), module will be null, and ProcessModule will be constructed wrapping that null value.  ProcessModule.ModuleName looks like this:
```
public string ModuleName
{
    get { return _moduleInfo._baseName; }
}
```
so if the ProcessModule is constructed to wrap a null value, _moduleInfo will be null and this will dereference a null reference.

In other words, there's a real product bug of some sort here, and even if updating this test hides that, we shouldn't lose sight of there being an actual issue to fix.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 9]
```diff
@@ -3,580 +3,381 @@
...
+        }
 
-            s_Process_EnableRaiseEvents_isExitedEventHandlerCalled = false;
+        [Fact]
+        public void Process_EnableRaiseEvents()
+        {
+            {
+                // Test behavior when EnableRaisingEvent = true;
+                // Ensure event is called.
+                s_Process_EnableRaiseEvents_isExitedEventHandlerCalled = false;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Why is this global state necessary?  Couldn't we just hook up a lambda to p.Exited that accesses a local?  I realize you didn't introduce this field, but as long as we're cleaning up the tests, it'd be nice to clean this up as well.  Things like this will make it easier for bugs to slip in.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 10]
```diff
@@ -3,580 +3,381 @@
...
+        public void Process_ExitCode()
         {
-            Process p = CreateProcess();
-            p.Start();
-            p.WaitForExit();
-            Assert.Equal(p.ExitCode, 100);
+            {
+                Process p = CreateProcess();
+                p.Start();
+                DelayTask();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Why are we not waiting for exit here?  Seems like we're relying on timing to ensure that the process has exited.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 11]
```diff
@@ -3,580 +3,381 @@
...
-            Assert.True(p.HasExited, "Process_HasExited001 failed");
-
-            p = CreateProcessInfinite();
-            p.Start();
-            try
             {
-                Assert.False(p.HasExited, "Process_HasExited002 failed");
+                Process p = CreateProcess();
+                p.Start();
+                DelayTask();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Same wait for exit question here.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 12]
```diff
@@ -3,580 +3,381 @@
...
-            var text = new StringBuilder();
-            text.AppendLine("Modules Collection:");
-            foreach (ProcessModule module in modules)
-            {
-                text.Append("    ");
-                text.AppendLine(GetModuleDescription(module));
-            }
-            return text.ToString();
+            int intCurrValue = (Int32)s_process.MaxWorkingSet;
+            s_process.MaxWorkingSet = (IntPtr)(intCurrValue + 1024);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Could this overflow?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 13]
```diff
@@ -3,580 +3,381 @@
...
-            {
-                int x = p.SessionId;
-            }
-            finally
-            {
-                p.Kill();
-                p.WaitForExit();
-            }
+            IntPtr curProcessorAffinity = s_process.ProcessorAffinity;
+            s_process.ProcessorAffinity = new IntPtr(0x1);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Do we want a try/finally to reset the affinity in case of failure?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 14]
```diff
@@ -3,580 +3,381 @@
...
-            }
+            bool isPriorityBoostEnabled = s_process.PriorityBoostEnabled;
+
+            s_process.PriorityBoostEnabled = true;
+            Assert.True(s_process.PriorityBoostEnabled, "Process_PriorityBoostEnabled001 failed");
+
+            s_process.PriorityBoostEnabled = false;
+            Assert.False(s_process.PriorityBoostEnabled, "Process_PriorityBoostEnabled002 failed");
+
+            s_process.PriorityBoostEnabled = isPriorityBoostEnabled;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Same finally question.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 15]
```diff
@@ -3,580 +3,381 @@
...
+
+            ProcessPriorityClass priorityClass = s_process.PriorityClass;
+
+            s_process.PriorityClass = ProcessPriorityClass.High;
+            Assert.Equal(s_process.PriorityClass, ProcessPriorityClass.High);
+
+            s_process.PriorityClass = ProcessPriorityClass.Normal;
+            Assert.Equal(s_process.PriorityClass, ProcessPriorityClass.Normal);
+
+            s_process.PriorityClass = priorityClass;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Ditto.  I'll stop asking :)

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 16]
```diff
@@ -3,580 +3,381 @@
...
-            try
-            {
-                Assert.Equal(p.ProcessName, "ProcessTest_ConsoleApp");
-            }
-            finally
-            {
-                p.Kill();
-                p.WaitForExit();
-            }
+            Assert.Equal(s_process.ProcessName, "ProcessTest_ConsoleApp");
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: We stored a process name in a constant earlier.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 17]
```diff
@@ -3,580 +3,381 @@
...
         [Fact]
-        public static void Process_GetProcessById()
+        public void Process_GetCurrentProcess()
         {
-            Process p = Process.GetCurrentProcess();
-            Process p_copy = Process.GetProcessById(p.Id);
-            Assert.Equal(p.ProcessName, p_copy.ProcessName);
-            Assert.Equal(p.Id, p_copy.Id);
+            int currentProcessId = ProcessTest.GetCurrentProcessId();
+            Assert.NotSame(0, currentProcessId);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: NotEqual?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 18]
```diff
@@ -3,580 +3,381 @@
...
+            Assert.Equal(currentProcessId, Process.GetCurrentProcess().Id);
+            Assert.Equal(Process.GetProcessById(currentProcessId).ProcessName, Process.GetCurrentProcess().ProcessName);
         }
 
         [Fact]
-        public static void Process_GetProcesses()
+        public void Process_GetProcesses()
         {
             // Get all the processes running on the machine.
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: The comment is incorrect.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs [hunk 19]
```diff
@@ -9,97 +9,103 @@ namespace System.Diagnostics.ProcessTests
...
-            if (!p.HasExited)
-            {
-                string s = p.StandardError.ReadToEnd();
-                p.WaitForExit();
-
-                s = s.Replace("\r", "").Replace("\n", "");
-
-                Assert.Equal("Unhandled Exception: Intentional Exception thrown", s);
-            }
+            Assert.Equal(p.StandardError.ReadToEnd(), "ProcessTest_ConsoleApp.exe error stream\r\n");
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs]: Shouldn't we wait/kill the process in a finally?

File: src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs (hunk 2 above)
Comment (pallavit) [inline src/System.Diagnostics.Process/tests/ProcessTest_ConsoleApp/ProcessTest_ConsoleApp.cs]: We are adding the delay to pretend the process is doing some work, to ensure that the process is active until WaitForExit() is executed.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 3 above)
Comment (pallavit) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: CreateProcessInfinite() is explicitly killed much before that. Personally I liked Thread.Sleep(-1) and ProcessTest.Dispose() method combination however this was explicitly replaced at one point to be 30 secs so I let it stay.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 4 above)
Comment (pallavit) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Taken care.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 8 above)
Comment (pallavit) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: I do not intend to hide the issue and will create a separate tracking bug for the fix. However I would like to keep are testing consistent.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs (hunk 19 above)
Comment (pallavit) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs]: I have added WaitForExit().

**Comment by stephentoub** [comment]
Thanks for making those changes, Pallavi.  I've left a few more comments.

**Comment by krwq** [comment]
@pallavit, please fix the merge conflict

**Comment by pallavit** [comment]
Incorporated all the comments but closing the PR and recreating from the private repo.
